packages/aws-cdk-lib/aws-dynamodb/lib/table.ts

```diff
@@ -370,6 +370,13 @@ export interface TableOptions extends SchemaOptions {
    * @default - no data import from the S3 bucket
    */
   readonly importSource?: ImportSourceSpecification;
+
+  /**
+   * Resource policy to assign to DynamoDB Table.
+   *
+   * @default - No resource policy statements are added to the created table.
+   */
+  readonly resourcePolicy?: iam.PolicyDocument;
```

Comment:
Don't expose the `resourcePolicy` like this, integrate it with the L2's [grant](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-dynamodb/lib/table.ts#L515) methods (try using `addToPrincipalOrResource`). You'll need to put this under a feature flag, because otherwise this will be modifying people's IAM permissions on a version bump, which is a breaking change.

Reply:
This is the same suggestion that @toutou826 made here: https://github.com/aws/aws-cdk/pull/29818#issuecomment-2071127803.

Reply:
@comcalvi can you explain how it will work with the grant methods and why it would be backward incompatible? 

My understanding of `grant`, specifically `addToPrincipalOrResource` will first add IAM policy to the principal of the `grantee` resource and fall back to adding resource-based policy `addToResourcePolicy` on fail. For all existing grant usage for dynamodb table, it calls `addToPrincipal` which tries to add IAM policy and throw an exception when fail.

Are you suggesting to make `resourcePolicy` field private and `table` construct will need to implement `addToResourcePolicy` function? I don't see why this would be backward incompatible though.

Reply:
> Don't expose the `resourcePolicy` like this, integrate it with the L2's [grant](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-dynamodb/lib/table.ts#L515) methods (try using `addToPrincipalOrResource`). You'll need to put this under a feature flag, because otherwise this will be modifying people's IAM permissions on a version bump, which is a breaking change.

According to the [design guidelines](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#resource-policies) we should expose this prop:

>When a construct represents an AWS resource that supports a resource policy, it should expose an optional prop that will allow initializing resource with a specified policy [awslint:resource-policy-prop]:
```
resourcePolicy?: iam.PolicyStatement[]
```

Are you suggesting otherwise?

Reply:
I didn't realize this was in our design guidelines. We should follow the design guidelines here. As we discussed offline, this `grant` needs to support using `addToPrincpalOrResource`. `TableBase` needs to implement `IResourceWithPolicy`.